### PR TITLE
Switch to Monai Bundle config

### PR DESCRIPTION
**Pull request by ibro45**
Closes https://github.com/project-lighter/lighter/issues/7.

Still needs to be addressed:

- Monai Bundle Config doesn't support defining functions/methods using `_target_` (opened an issue https://github.com/Project-MONAI/MONAI/issues/5852 ) so I had to find a workaround. This enables passing the cli arguments to the Trainer method that you're running, e.g. `--fit#ckpt_path <PATH>`: https://github.com/project-lighter/lighter/blob/0e26608dedeecf81eb648ce942a140150880d72a/lighter/cli.py#L17
- Import project as a module does not work currently

Note:
- Calling `lighter fit --config_file ...` is basically equal to `python -m monai.bundle run fit --config_file ...`.

Conclusion:
- The configs remain basically the same apart from different referencing between Monai Bundle and Hydra/OmegaConf https://github.com/project-lighter/lighter/blob/monai-bundle/projects/cifar10/experiments/monai_bundle_prototype.yaml
- Allowed us to get rid of the `runner` and `config` modules

### CODE REVIEW

**Comment by ibro45** [comment]
@surajpaib Cannot seem to make the user project import work, even though the same implementation worked before monai bundle config

**Comment by surajpaib** [comment]
I think I understand what the issue is, the `project` function is only instantiated, but never called

**Comment by surajpaib** [comment]
Here is a working solution for your previous approach
``
``
project:
    _target_: lighter.utils.import_module_from_path
    module_name: project
    module_path: ./projects/cifar10

trainer:
    _target_: pytorch_lightning.Trainer
    max_epochs: 50
    accelerator: gpu
    devices: 1

system:
    _target_: lighter.LighterSystem
    batch_size: 512
    pin_memory: True
    num_workers: 4
    log_input_as: "image_batch"

    model:
        _target_: project.models.net.Net
        _requires_: "$@project()"
````

**Comment by surajpaib** [comment]
@ibro45 I'm thinking of a better way to do this overall, will post when I have an update

**Comment by ibro45** [comment]
Sorry, didn't see your comments. So, in the meantime i realized we don't need the `import project`. That makes it a bit better. In your example I have to say that I don't like having to call the partial function like `_requires_: "$@project()"`. Actually, I've been confused by why they didn't implement it so that it's called whenever you refer it, without all the $ and (). Let's see what happens with that PR they opened, I'll suggest it if they don't end up doing it.

**Comment by surajpaib** [comment]
I get why they did it this way though because you want a reference to the function and not call the function itself in many cases. And then be able to call it into your code however you want. 

For example. if you defined a train, test and val functions, and then want to call them accordingly. 
 I think its difficult to achieve both functionality together though :(

**Comment by ibro45** [comment]
Managed to get it to behave as it did before monai bundle config. Ready to be merged if you agree.

File: lighter/cli.py [hunk 1]
```diff
@@ -1,50 +1,102 @@
...
+from lighter.utils import import_module_from_path
 
 
-# Train
-@interface.command()
-@click.argument("args", nargs=-1)
-def train(args):
-    run("fit", args)
-
+def interface():
```
Comment (surajpaib) [inline lighter/cli.py:14]: I've refactored this bit as 

````
import sys
from functools import partial
from typing import Any, Dict

import fire
import yaml
from loguru import logger
from monai.bundle.scripts import run
from monai.utils.misc import ensure_tuple

from lighter.utils import import_module_from_path

# Trainer methods calls in YAML format with support for command line arguments.
# Waiting for https://github.com/Project-MONAI/MONAI/pull/5854#issuecomment-1384800886.
trainer_methods = {
    "fit": yaml.safe_load(
        """
        fit:
            _method_: >
                $@trainer.fit(model=@fit#model,
                              ckpt_path=@fit#ckpt_path)
            model: "@system"
            ckpt_path: null"""
    ),
    "validate": yaml.safe_load(
        """
        validate:
            _method_: > 
                $@trainer.validate(model=@validate#model,
                                   ckpt_path=@validate#ckpt_path,
                                   verbose=@validate#verbose)
            model: "@system"
            ckpt_path: null
            verbose: True"""
    ),
    "predict": yaml.safe_load(
        """
        predict:
            _method_: >
                $@trainer.predict(model=@predict#model,
                                  ckpt_path=@predict#ckpt_path)
            model: "@system"
            ckpt_path: null"""
    ),
    "test": yaml.safe_load(
        """
        test:
            _method_: >
                $@trainer.test(model=@test#model,
                               ckpt_path=@test#ckpt_path,
                               verbose=@test#verbose)
            model: "@system"
            ckpt_path: null
            verbose: True"""
    ),
    "tune": yaml.safe_load(
        """
        tune:
            _method_: > 
                $@trainer.tune(model=@tune#model,
                               ckpt_path=@tune#ckpt_path,
                               scale_batch_size_kwargs=@tune#scale_batch_size_kwargs,
                               lr_find_kwargs=@tune#lr_find_kwargs,
                               method=@tune#method)
            model: "@system"
            ckpt_path: null
            scale_batch_size_kwargs: null
            lr_find_kwargs: null
            method: fit
    """
    ),
}


def interface():
    fire.Fire({k: partial(run_trainer_method, v) for k, v in trainer_methods.items()})


def run_trainer_method(trainer_method: Dict, **kwargs: Any):
    """Call monai.bundle.run() on a Trainer method. If a project path
    is defined in the config file(s), import it.

    Args:
        trainer_method (Dict): Trainer method to run, collection defined in `trainer_methods`
        **kwargs (Any): keyword arguments for the monai.bundle.run function.
    """
    if "config_file" in kwargs:
        project_imported = False
        # Handle multiple configs
        for config in ensure_tuple(kwargs["config_file"]):
            config = yaml.safe_load(open(config))
            # Import the project as a module
            if "project" in config:
                # Only one config file can specify the project path
                if project_imported:
                    logger.error("`project` must be specified in one config only.")
                    sys.exit()
                import_module_from_path("project", config["project"])
                project_imported = True
    # Run the Trainer method
    run(**trainer_method, **kwargs)
````

What do you think @ibro45? I think the double dictionary definition  (in `trainer_methods` and `fire`) is slightly confusing. They are both the same thing, essentially.

**Review by surajpaib — COMMENTED** [review]
Amazing stuff with monai-bundle! Looks good to go, added a comment about a small refactor, let me know if you think it makes sense and I can add a commit to the PR

File: lighter/cli.py (hunk 1 above)
Comment (surajpaib) [reply lighter/cli.py:14]: I've referenced the `trainer_method` and ignored the `runner_id` here, but if we need to make it such that we have base defaults for a project then I'll add it again

Comment (ibro45) [reply lighter/cli.py:14]: I like it, thanks!

**Review by surajpaib — APPROVED** [review]
Looks good to go from me!

**Pull request merged**